From b9224b7ff648fec553d9f760d0ddbe4378bb0374 Mon Sep 17 00:00:00 2001
From: Kexy Biscuit <kexybiscuit@aosc.io>
Date: Sat, 17 Aug 2024 11:32:11 +0800
Subject: [PATCH 095/100] platform: x86: ideapad-laptop: add fixes for
 ThinkBook 14+ (2024 G6+)

This is a patch taken from ideapad-laptop-tb2024g6plus in an attempt to
incorporate these changes in the mainline kernel. This patch fixes
the following issues:

- Laptop shuts down unexpectedly after closing the lid.
- The Fn-F5 combination shuts down or resets AMD models.
- Keyboard backlight status resets after waking up from S3 suspend.

Co-authored-by: Terry <terry.wong2@yahoo.com>
Co-authored-by: Mingcong Bai <baimingcong@uniontech.com>
Ref: https://github.com/ty2/ideapad-laptop-tb2024g6plus
---
 drivers/platform/x86/ideapad-laptop.c | 29 ++++++++++++++++++++++++---
 1 file changed, 26 insertions(+), 3 deletions(-)

diff --git a/drivers/platform/x86/ideapad-laptop.c b/drivers/platform/x86/ideapad-laptop.c
index 98ec30fce9fdd..a3b57ddb41b80 100644
--- a/drivers/platform/x86/ideapad-laptop.c
+++ b/drivers/platform/x86/ideapad-laptop.c
@@ -160,6 +160,7 @@ struct ideapad_private {
 		struct led_classdev led;
 		unsigned int last_brightness;
 	} fn_lock;
+	bool suspended;
 };
 
 static bool no_bt_rfkill;
@@ -1153,6 +1154,8 @@ static const struct key_entry ideapad_keymap[] = {
 	{ KE_KEY,	0x27 | IDEAPAD_WMI_KEY, { KEY_HELP } },
 	/* Refresh Rate Toggle */
 	{ KE_KEY,	0x0a | IDEAPAD_WMI_KEY, { KEY_REFRESH_RATE_TOGGLE } },
+	/* Touchpad Toggle */
+	{ KE_KEY,  0x29 | IDEAPAD_WMI_KEY, { KEY_TOUCHPAD_TOGGLE } },
 
 	{ KE_END },
 };
@@ -1702,6 +1705,16 @@ static void ideapad_acpi_notify(acpi_handle handle, u32 event, void *data)
 	unsigned long vpc1, vpc2, bit;
 
 	scoped_guard(mutex, &priv->vpc_mutex) {
+		acpi_handle_info(handle, "event: %lu\n",
+						(unsigned long)event);
+
+		if(!data)
+			acpi_handle_info(handle, "no data");
+			return;
+
+		if (priv->suspended)
+			return;
+
 		if (read_ec_data(handle, VPCCMD_R_VPC1, &vpc1))
 			return;
 
@@ -1748,6 +1761,7 @@ static void ideapad_acpi_notify(acpi_handle handle, u32 event, void *data)
 			ideapad_backlight_notify_power(priv);
 			break;
 		case KBD_BL_KBLC_CHANGED_EVENT:
+		case 0:
 		case 1:
 			/*
 			 * Some IdeaPads report event 1 every ~20
@@ -1757,8 +1771,6 @@ static void ideapad_acpi_notify(acpi_handle handle, u32 event, void *data)
 			 * backlight has changed.
 			 */
 			ideapad_kbd_bl_notify(priv);
-			break;
-		case 0:
 			ideapad_check_special_buttons(priv);
 			break;
 		default:
@@ -1962,6 +1974,8 @@ static const struct wmi_device_id ideapad_wmi_ids[] = {
 	{ "26CAB2E5-5CF1-46AE-AAC3-4A12B6BA50E6", &ideapad_wmi_context_esc }, /* Yoga 3 */
 	{ "56322276-8493-4CE8-A783-98C991274F5E", &ideapad_wmi_context_esc }, /* Yoga 700 */
 	{ "8FC0DE0C-B4E4-43FD-B0F3-8871711C1294", &ideapad_wmi_context_fn_keys }, /* Legion 5 */
+	{ "46f16367-fb9d-11ee-a4f6-40c2ba4a5625", &ideapad_wmi_context_esc }, /* ThinkBook 16+ 2024 IMH */
+	{ "077c4a1f-e344-11ee-a4f6-40c2ba413e67", &ideapad_wmi_context_esc }, /* ThinkBook 2024 AMD */
 	{},
 };
 MODULE_DEVICE_TABLE(wmi, ideapad_wmi_ids);
@@ -2152,10 +2166,19 @@ static int ideapad_acpi_resume(struct device *dev)
 	if (priv->dytc)
 		dytc_profile_refresh(priv);
 
+	priv->suspended = false;
+
+	return 0;
+}
+
+static int ideapad_acpi_suspended(struct device *dev)
+{
+	struct ideapad_private *priv = dev_get_drvdata(dev);
+	priv->suspended = true;
 	return 0;
 }
 #endif
-static SIMPLE_DEV_PM_OPS(ideapad_pm, NULL, ideapad_acpi_resume);
+static SIMPLE_DEV_PM_OPS(ideapad_pm, ideapad_acpi_suspended, ideapad_acpi_resume);
 
 static const struct acpi_device_id ideapad_device_ids[] = {
 	{"VPC2004", 0},
-- 
2.46.0

